WalletWasabi/WabiSabi/Models/DecompositionAlgs/Decomposition.cs

```diff
@@ -0,0 +1,59 @@
+using NBitcoin;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Models.DecompositionAlgs
+{
+	public class Decomposition : IEnumerable<Coin>
+	{
+		private List<Coin> Coins { get; } = new();
+
+		public void Extend(Coin coin)
+		{
+			int index = Coins.BinarySearch(coin);
+			if (index < 0)
+			{
+				Coins.Insert(~index, coin);
+			}
+		}
+
+		public Money FaceValue()
+		{
+			return Coins.Sum(c => c.Amount);
+		}
+
+		public int OutputsVbytes()
+		{
+			return Coins.Sum(_ => Coin.OutputVbytes);
+		}
+
+		public decimal InputsVbytes()
+		{
+			return Coins.Sum(_ => Coin.InputVbytes);
+		}
+
+		public Money EffectiveValue(FeeRate feeRate)
+		{
+			var sats = (long)Math.Floor(feeRate.SatoshiPerByte * Coin.InputVbytes * Coins.Count);
+			return FaceValue() - Money.Satoshis(sats);
+		}
+
+		public Money EffectiveCost(FeeRate feeRate)
+		{
+			var sats = (long)Math.Floor(feeRate.SatoshiPerByte * Coin.OutputVbytes * Coins.Count);
+			return FaceValue() + Money.Satoshis(sats);
+		}
+
+		public IEnumerator<Coin> GetEnumerator()
+		{
+			return Coins.GetEnumerator();
+		}
+
+		IEnumerator IEnumerable.GetEnumerator()
+		{
+			return Coins.GetEnumerator();
+		}
```

Comment:
We can simply expose a property `Coins => theListOfCoins.ToImmutableArray()` and remove this methods.

Reply:
can we just subclass `ImmutableList<Coin>`?

Reply:
no possible: sealed class `ImmutableList`

Reply:
Forget about Immutables. We were fine without them. No need to further complicate the code because of this. As I said before, we are not using it in the codebase.

Reply:
If we want to change this. we can do it later after ww2.

Reply:
In this case it really matters because if we use `List` we will have to clone them when we compose decomposers, it's not a style issue but how `Add` works.

The correct abstraction for a decomposition is an immutable list because it represents a point in the search space and should not be modifiable when these points are manipulated by different components.

Since we don't need any methods we can just remove `Decomposition` and use `ImmutableList<Money>` without subclassing it, and if we ever need that concept to come back we can wrap it.

Reply:
I like the remove `Decomposition` part.